providers/implementations/digests/sha3_prov.c

```diff
@@ -249,6 +249,65 @@ static PROV_SHA3_METHOD kmac_s390x_md =
     } else {                                                                   \
         ctx->meth = sha3_generic_md;                                           \
     }
+#elif defined(__aarch64__)
+# include "arm_arch.h"
+
+static sha3_absorb_fn armsha3_sha3_absorb;
+
+size_t SHA3_absorb_cext(uint64_t A[5][5], const unsigned char *inp, size_t len,
+                    size_t r);
+/*-
+ * Hardware-assisted ARMv8.2 SHA3 extension version of the absorb()
+ */
+static size_t armsha3_sha3_absorb(void *vctx, const void *inp, size_t len)
+{
+    KECCAK1600_CTX *ctx = vctx;
+
+    return SHA3_absorb_cext(ctx->A, inp, len, ctx->block_size);
+}
+
+static PROV_SHA3_METHOD sha3_ARMSHA3_md =
+{
+    armsha3_sha3_absorb,
+    generic_sha3_final
+};
+/* Detection on Apple operating systems */
+# if defined(__APPLE__)
+#  define ARM_SHA3_CAPABLE (OPENSSL_armcap_P & ARMV8_SHA3)
+#  define SHA3_SET_MD(uname, typ)                                              \
+    if (ARM_SHA3_CAPABLE) {                                                    \
+        ctx->meth = sha3_ARMSHA3_md;                                           \
+    } else {                                                                   \
+        ctx->meth = sha3_generic_md;                                           \
+    }
+#  define KMAC_SET_MD(bitlen)                                                  \
+    if (ARM_SHA3_CAPABLE) {                                                    \
+        ctx->meth = sha3_ARMSHA3_md;                                           \
+    } else {                                                                   \
+        ctx->meth = sha3_generic_md;                                           \
+    }
+/* Detection on other operating systems */
+# else
+#  define ARM_HAS_FASTER_SHA3                                                                  \
+    (MIDR_IS_CPU_MODEL(OPENSSL_arm_midr, ARM_CPU_IMP_APPLE, APPLE_CPU_PART_M1_FIRESTORM)     ||\
+     MIDR_IS_CPU_MODEL(OPENSSL_arm_midr, ARM_CPU_IMP_APPLE, APPLE_CPU_PART_M1_FIRESTORM_PRO) ||\
+     MIDR_IS_CPU_MODEL(OPENSSL_arm_midr, ARM_CPU_IMP_APPLE, APPLE_CPU_PART_M1_FIRESTORM_MAX) ||\
+     MIDR_IS_CPU_MODEL(OPENSSL_arm_midr, ARM_CPU_IMP_APPLE, APPLE_CPU_PART_M2_AVALANCHE)     ||\
+     MIDR_IS_CPU_MODEL(OPENSSL_arm_midr, ARM_CPU_IMP_APPLE, APPLE_CPU_PART_M2_AVALANCHE_PRO) ||\
+     MIDR_IS_CPU_MODEL(OPENSSL_arm_midr, ARM_CPU_IMP_APPLE, APPLE_CPU_PART_M2_AVALANCHE_MAX))
```

Comment:
Can you use `sysctl_query()` instead, looking for features, rather than having to hard-code a long list of part numbers?

Look in `crypto/armcap.c`, where this is usually done

Reply:
Is `sysctl_query()` available on Linux?

Reply:
Look at how it's done in `crypto/armcap.c`, which abstracts all this away - and yes, it is done differently on Linux, macOS, Windows, etc

Reply:
(Sorry my reply via email was put in the wrong place)

> Look at how it's done for other optimisations, such as ARMV8_UNROLL8_EOR3: the decision is taken in armcap.c, and the implementation only looks at OPENSSL_armcap_P to know whether it can use a particular optimisation or not.

I just noticed your discussions in #15916. I feel the same as @bernd-edlinger 's [first impression](https://github.com/openssl/openssl/pull/15916#discussion_r711238403). Here will be stranger, as there is already an `ARMV8_SHA3` in `OPENSSL_armcap_P`, and we are going to define another `ARMV8_SHA3_INDEED_FASTER` or whatever?

> This makes it simple to test optimisations on new hardware: just set OPENSSL_armcap_P in the process environment and run openssl speed.

Yes, `OPENSSL_armcap_P` is very convenient for selecting code paths. But as there are potentially many combinations of algorithms and CPU microarchitectures, if a new bit is assigned for every cases, will this make `OPENSSL_armcap_P` bloated and users confused?

And for my code, `OPENSSL_armcap` environment variable indeed works for `__APPLE__` by clearing the `ARMV8_SHA3` bit. And this function can be added for the non `__APPLE__` case by adding a check for `ARM_SHA3_CAPABLE` in `ARM_HAS_FASTER_SHA3` macro.

> It also means that when new CPU microarchitectures are released, with different MIDR values, there's only one place in the code to change, instead of having to grep throughout all the algorithms to see which ones might need tweaking.

Maybe move the `ARM_HAS_FASTER_SHA3` macro into `crypto/arm_arch.h`? But again this will make the macro globally visible. Users may be confused as this `ARM_HAS_FASTER_SHA3` is only related to `keccak1600-armv8.pl`.

> And also note that MIDR, unlike Intel's CPUID, is not readable by user code (the official name is MIDR_EL1, which means it can only be read at EL1 or above, which is the kernel). This allows the kernel to handle differences in CPUs on heterogenous systems (e.g. big.LITTLE where the different CPU types aren't compatible).

Thanks for the information! I just use the openssl initialized `OPENSSL_arm_midr` variable and on Linux it has the correct value. It seems Linux provides an emulation mechanism to read the register (#11744).

And maybe in the future `OPENSSL_arm_midr` can also be overridden by environment variable like `OPENSSL_armcap` to fake the CPU model for testing? I think this is better than putting all CPU information and algorithm combinations into `OPENSSL_armcap` bits.